File: packages/runtime/src/path.ts
export namespace path {
	export type Component =
		| Component.Current
		| Component.Normal
		| Component.Parent
		| Component.Root;

	export namespace Component {
		export type Current = ".";

		export let Current = ".";

		export type Normal = string;

		export type Parent = "..";

		export let Parent = "..";

		export type Root = "/";

		export let Root = "/";

		export let isNormal = (component: Component): component is Normal => {
			return (
				component !== Current && component !== Parent && component !== Root
			);
		};
	}

	export let components = (arg: string): Array<Component> => {
		let components = arg.split("/");
		if (components.at(0)?.length === 0) {
			components[0] = Component.Root;
		}
		components = components.filter((component, i) => {
			if (i > 0 && component === Component.Current) {
				return false;
			}
			if (component.length === 0) {
				return false;
			}
			return true;
		});
		return components;
	};

	export let fromComponents = (components: Array<Component>): string => {
		if (components[0] === Component.Root) {
			return `/${components.slice(1).join("/")}`;
		} else {
			return components.join("/");
		}
	};

	export let isAbsolute = (arg: string): boolean => {
		return arg.startsWith("/");
	};

	export let join = (...args: Array<string | undefined>): string => {
		let components: Array<string> = [];
		for (let arg of args) {
			if (arg === undefined) {
				continue;
			}
			if (isAbsolute(arg)) {
				components = path.components(arg);
			} else {
				components = components.concat(path.components(arg));
			}
		}
		return fromComponents(components);
	};

	export let parent = (arg: string): string | undefined => {
		let components = path.components(arg);
		if (components.length === 0) {
			return undefined;
		}
		return components.slice(0, -1).join("/");
	};
}

File: packages/clients/python/tangram/path.py
"""Path utilities matching JS path.ts."""

from typing import List, Optional, Union

# Component types.
CURRENT = "."
PARENT = ".."
ROOT = "/"

Component = str


def is_normal(component: Component) -> bool:
    """Check if a component is a normal path segment."""
    return component not in (CURRENT, PARENT, ROOT)


def components(arg: str) -> List[Component]:
    """Split a path into its components."""
    parts = arg.split("/")
    result: List[Component] = []

    # Handle root.
    if parts and len(parts[0]) == 0:
        result.append(ROOT)
        parts = parts[1:]

    for i, part in enumerate(parts):
        # Skip empty parts.
        if len(part) == 0:
            continue
        # Skip . except at the beginning.
        if part == CURRENT and i > 0:
            continue
        result.append(part)

    return result


def from_components(comps: List[Component]) -> str:
    """Join components back into a path string."""
    if comps and comps[0] == ROOT:
        return "/" + "/".join(comps[1:])
    return "/".join(comps)


def is_absolute(arg: str) -> bool:
    """Check if a path is absolute."""
    return arg.startswith("/")


def join(*args: Optional[str]) -> str:
    """Join path segments together."""
    result: List[Component] = []

    for arg in args:
        if arg is None:
            continue
        if is_absolute(arg):
            result = components(arg)
        else:
            result = result + components(arg)

    return from_components(result)


def parent(arg: str) -> Optional[str]:
    """Get the parent directory of a path."""
    comps = components(arg)
    if not comps:
        return None
    return "/".join(comps[:-1]) if comps[:-1] else None


def basename(arg: str) -> str:
    """Get the last component of a path."""
    comps = components(arg)
    if not comps:
        return ""
    last = comps[-1]
    return "" if last == ROOT else last


def dirname(arg: str) -> str:
    """Get the directory name of a path."""
    p = parent(arg)
    return p if p is not None else ""


def extname(arg: str) -> str:
    """Get the file extension of a path."""
    base = basename(arg)
    if "." in base:
        return "." + base.rsplit(".", 1)[-1]
    return ""


def normalize(arg: str) -> str:
    """Normalize a path by resolving . and .. components."""
    comps = components(arg)
    result: List[Component] = []
    is_abs = comps and comps[0] == ROOT

    for comp in comps:
        if comp == ROOT:
            result = [ROOT]
        elif comp == CURRENT:
            continue
        elif comp == PARENT:
            if result and result[-1] not in (ROOT, PARENT):
                result.pop()
            elif not is_abs:
                result.append(PARENT)
        else:
            result.append(comp)

    return from_components(result)
